packages/@sanity/core/src/actions/graphql/extractFromSanitySchema.js

```diff
@@ -0,0 +1,410 @@
+    candidates.forEach((def, i) => {
+      if (typeNeedsHoisting(def)) {
+        throw createLiftTypeArrayError(
+          i,
+          parent.name,
+          def.type ? def.type.name : def.name,
+          options.grandParent
+        )
+      }
+    })
+
+    const converted = candidates.map(def => convertType(def))
+
+    // We might end up with union types being returned - these needs to be flattened
+    // so that an ImageOr(PersonOrPet) becomes ImageOrPersonOrPet
+    const flattened = converted.reduce((acc, candidate) => {
+      const union = unionTypes.find(item => item.name === candidate.type)
+      return union
+        ? acc.concat(union.types.map(type => ({type, isReference: candidate.isReference})))
+        : acc.concat(candidate)
+    }, [])
+
+    const allCandidatesAreDocuments = flattened.every(def => {
+      const typeDef = sanityTypes.find(type => type.name === (def.type.name || def.type))
+      return typeDef && typeDef.type === 'document'
+    })
+
+    const interfaces = allCandidatesAreDocuments ? ['Document'] : undefined
+    const refs = flattened.filter(type => type.isReference).map(ref => ref.type)
+    const possibleTypes = flattened.map(type => (type.isReference ? type.type : type.name)).sort()
+
+    const name = possibleTypes.join('Or')
+
+    if (!unionTypes.some(item => item.name === name)) {
+      unionTypes.push({
+        kind: 'Union',
+        name,
+        types: possibleTypes,
+        interfaces
+      })
+    }
+
+    const references = refs.length > 0 ? refs : undefined
+    return {type: name, references}
+  }
+
+  function getDocumentDefinition(def) {
+    const objectDef = getObjectDefinition(def)
+    const fields = getDocumentInterfaceFields().concat(objectDef.fields)
+
+    return Object.assign(objectDef, {
+      fields,
+      interfaces: ['Document']
+    })
+  }
+
+  function getDocumentInterfaceDefinition() {
+    return {
+      kind: 'Interface',
+      name: 'Document',
+      description: 'A Sanity document',
+      fields: getDocumentInterfaceFields()
+    }
+  }
+
+  function getDocumentInterfaceFields() {
+    return [
+      {
+        fieldName: '_id',
+        type: 'ID',
+        isNullable: false,
+        description: 'Document ID'
+      },
+      {
+        fieldName: '_type',
+        type: 'String',
+        isNullable: false,
+        description: 'Document type'
+      },
+      {
+        fieldName: '_createdAt',
+        type: 'Datetime',
+        isNullable: false,
+        description: 'Date the document was created'
+      },
+      {
+        fieldName: '_updatedAt',
+        type: 'Datetime',
+        isNullable: false,
+        description: 'Date the document was last modified'
+      },
+      {
+        fieldName: '_rev',
+        type: 'String',
+        isNullable: false,
+        description: 'Current document revision'
+      }
+    ]
+  }
+
+  function arrayify(thing) {
+    return Array.isArray(thing) ? thing : [thing]
```

Comment:
Depending on how `arrayify` is called, this could potentially yield `[undefined]` if `thing` is `undefined`. Maybe do an additional check and return an empty array in that case.